### PR TITLE
Exclude string.ToLowerInvariant() in GetEnvironmentVariableAsBool()

### PR DESCRIPTION
**Pull request by iSazonov**
# PR Summary

Exclude string.ToLowerInvariant() in GetEnvironmentVariableAsBool() to avoid early ICU initialization.

## PR Context

Related #14268.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [x] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [ ] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

### CODE REVIEW

**Comment by daxian-dbw** [comment]
@iSazonov I doubt the usefulness of this change. The most comment scenario would be the `POWERSHELL_TELEMETRY_OPTOUT` environment variable is not defined. The `GetEnvironmentVariableAsBool` method would just return from the `if (string.IsNullOrEmpty(str))` block, never hitting the `switch (str.ToLowerInvariant())` line.

Your changes here has zero impact to the most common scenario. It improves a less common scenario with the cost of decreased readability, so I'm not convinced we want it. @JamesWTruher, can you please review and share your thoughts?

**Comment by iSazonov** [comment]
@daxian-dbw 
As we already know, ApplicationInsights affects the startup speed unpredictably. So I turned off the telemetry locally to explore other areas.
The funny thing is that this code is the first thing I saw in the trace. It's a hindrance again. 
Ideally we could use bool.Parse() method but it only support "true"/"false" strings.

**Comment by daxian-dbw** [comment]
@JamesWTruher Additional context info: `string.ToLowerInvariant()` triggers the loading of the International Component of Unicode package, which takes ~11ms in my local measurement.

**Comment by JamesWTruher** [comment]
I'm somewhat in agreement with @daxian-dbw here. I don't see _any_ actual measurement of performance change. If this is about improving performance, I would love to see what this actually gets us.

**Comment by iSazonov** [comment]
I would like to avoid early loading of ICU. This is how it might look:
image

**Review by daxian-dbw — CHANGES_REQUESTED** [review]
@iSazonov I understand it's desired to avoid `ToLowerInvariant`. I think the concern is, the change improves a less common code path with the cost in readability. I will talk with @JamesWTruher more about it.

File: src/System.Management.Automation/utils/Telemetry.cs [hunk 1]
```diff
@@ -563,19 +563,56 @@ private static bool GetEnvironmentVariableAsBool(string name, bool defaultValue)
                 return defaultValue;
             }
 
-            switch (str.ToLowerInvariant())
+            var boolStr = str.AsSpan().Trim();
```
[the remark this answers is not in the document]
Comment (iSazonov) [reply src/System.Management.Automation/utils/Telemetry.cs:566]: This follow the intention to support many acceptable values.
But I wonder why we need this at all. We could support only "0" and "1" values or better to check only presents the variable as a flag.

[the remark this answers is not in the document]
Comment (daxian-dbw) [reply src/System.Management.Automation/utils/Telemetry.cs:566]: Don't change behavior when doing perf optimization. Also, we cannot do breaking changes.

[the remark this answers is not in the document]
Comment (iSazonov) [reply src/System.Management.Automation/utils/Telemetry.cs:566]: I removed the Trim().
Code below comes from .Net https://source.dot.net/#System.Private.CoreLib/Boolean.cs,198ff42f14d8c64b

I could make the method internal and add xUnit test for the method.

[the remark this answers is not in the document]
Comment (daxian-dbw) [reply src/System.Management.Automation/utils/Telemetry.cs:566]: > I removed the Trim().

@iSazonov Can you push your latest change?

[the remark this answers is not in the document]
Comment (iSazonov) [reply src/System.Management.Automation/utils/Telemetry.cs]: Oh, sorry - forgot to push the commit.

File: src/System.Management.Automation/utils/Telemetry.cs [hunk 2]
```diff
@@ -563,19 +563,56 @@ private static bool GetEnvironmentVariableAsBool(string name, bool defaultValue)
                 return defaultValue;
             }
 
-            switch (str.ToLowerInvariant())
+            var boolStr = str.AsSpan();
```
Comment (daxian-dbw) [inline src/System.Management.Automation/utils/Telemetry.cs:566]: It looks to me `AsSpan` is not needed. You can just use `str` for the subsequent character comparisons.

**Review by daxian-dbw — COMMENTED** [review]
@iSazonov Sorry for being sluggish. We can merge this one after you address the one more comment I left.

File: src/System.Management.Automation/utils/Telemetry.cs (hunk 2 above)
Comment (SeeminglyScience) [reply src/System.Management.Automation/utils/Telemetry.cs:566]: > It looks to me `AsSpan` is not needed. You can just use `str` for the subsequent character comparisons.

JIT optimization is a little bit [more favorable with span](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIGYACMhgYQYG8aHunHjykDYBAgAbBgEFcAZQAO2AHYAKfgAYGuDFDRCR4gCYwAZtgCuojADVso0zACUHLjxcA3bFF1jpWhgF4NLQA6KTlFJXsAbhpnF24ASyMGJWFvYIAZGAUAcwwAC38A8ntYuM5qOMqGROTU0R8oAG1VAF1ChgBycg6SiqqXcv7+4gB2Bi07aL6hhgBfGOmhmpS9Bua2vwCO1R7S/sGZuNGGE1FcGCnDub2eeeobhKSVtKggzJz89sYAMm+HuOe9S063aHQAnh0GIRCF4gU1WqCAJo9Bi/f4uQFrcgbLYwSHQ2FYnGdACiKLRi36mOBpGJHVw+JhdTWtNB0h6vX26J4xwmF3+d3+x0MJnMVhsk1KdxcpVK9CYAlhkhkWniORU5HUmm0SpFZgs1lsDiclJ47k8zN8AW1U3+y0tr3euQKmwYxX+Bxm9tWwIRrq6u1NlU9VyYYz5l0OgqDcW9LxB/p2nMOIaux1O50jM2lVWj/TjcLeWWdX1RfxjGIdCa2EKhTJ98LpyMcFKu1Ka2NBeLrhOBnf9ZJb5bbVdZ/oZPdHdPZ9mTwe53F5UElMbzVWFxn14qNWeu1FmQA=) due to eliding bounds checks.  Unless that was also added to string and just hasn't hit sharplab yet.

(Not making a statement of worth, just making sure the difference is known)

Comment (daxian-dbw) [reply src/System.Management.Automation/utils/Telemetry.cs:566]: Indeed. Bounds check is removed for the `span` case once you start the check with `boolStr.Length == ...`, but not for the `string` case. This is good to know, thanks @SeeminglyScience.

Comment (iSazonov) [reply src/System.Management.Automation/utils/Telemetry.cs:566]: Closed?

**Comment by iSazonov** [comment]
> Sorry for being sluggish.

Sometimes holidays take more energy than work :laughing:

**Review by daxian-dbw — APPROVED** [review]

**Pull request merged**